pulsar-jms/src/main/java/com/datastax/oss/pulsar/jms/MessagePriorityGrowableArrayBlockingQueue.java

```diff
@@ -20,17 +20,68 @@
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
 import org.apache.pulsar.client.api.Message;
 import org.apache.pulsar.common.util.collections.GrowableArrayBlockingQueue;
 
+import javax.annotation.security.RunAs;
+
+@Slf4j
 public class MessagePriorityGrowableArrayBlockingQueue extends GrowableArrayBlockingQueue<Message> {
 
+  @AllArgsConstructor
+  static class StaleMessagesChecker implements Runnable {
+    private final MessagePriorityGrowableArrayBlockingQueue queue;
+    private final org.apache.pulsar.client.api.Consumer consumer;
+
+    @Override
+    public void run() {
+      try {
+        long now = System.currentTimeMillis();
+        long expireBefore = now - TimeUnit.SECONDS.toMillis(10);
+        List<MessageInfo> removed = new ArrayList<>();
+        queue.queue.forEach((info) -> {
+          if (info.insertTs <= expireBefore) {
+            removed.add(info);
+          }
+        });
+        queue.queue.removeAll(removed);
+        for (MessageInfo messageInfo : removed) {
+
+          try {
+            // this operation shouldn't fail because we're simply adding it to the nack backlog
+            // but it's better to be conservative
+            consumer.negativeAcknowledge(messageInfo.message);
```

Comment:
IIRC negativeAcknowledge is not available for all the subscription types (only Shared?)